cmd/snap-repair/runner.go

```diff
@@ -43,6 +47,138 @@ import (
 	"github.com/snapcore/snapd/strutil"
 )
 
+// Repair is a runnable repair.
+type Repair struct {
+	*asserts.Repair
+
+	run      *Runner
+	sequence int
+}
+
+func (r *Repair) RunDir() string {
+	return filepath.Join(dirs.SnapRepairRunDir, r.BrandID(), r.RepairID())
+}
+
+// SetStatus sets the status of the repair in the state and saves the latter.
+func (r *Repair) SetStatus(status RepairStatus) {
+	brandID := r.BrandID()
+	cur := *r.run.state.Sequences[brandID][r.sequence-1]
+	cur.Status = status
+	r.run.setRepairState(brandID, cur)
+	r.run.SaveState()
+}
+
+// nextCounter finds the next counter prefix to use for the output and status
+// files of a repair run
+func nextCounter(r *Repair) int {
+	counter := 0
+
+	// format: "r99.000001.2006-01-02T150405.999999999.output"
+	prefix := fmt.Sprintf("r%d.", r.Revision())
+	suffix := fmt.Sprintf(".output")
+	li, err := ioutil.ReadDir(r.RunDir())
+	if err != nil {
+		// should never happen
+		return counter
+	}
+	for _, dirent := range li {
+		name := dirent.Name()
+		if strings.HasPrefix(name, prefix) && strings.HasSuffix(name, suffix) {
+			li := strings.Split(name, ".")
+			if c, err := strconv.Atoi(li[1]); err == nil {
+				counter = c
+			}
+		}
+	}
+
+	return counter + 1
+}
+
+// Run executes the repair script leaving execution trail files on disk.
+func (r *Repair) Run() error {
+	// write the script to disk
+	rundir := r.RunDir()
+	err := os.MkdirAll(rundir, 0775)
+	if err != nil {
+		return err
+	}
+
+	script := filepath.Join(rundir, fmt.Sprintf("script.r%d", r.Revision()))
+	err = osutil.AtomicWriteFile(script, r.Body(), 0700, 0)
+	if err != nil {
+		return err
+	}
+
+	// the date may be broken so we use an additional counter
+	counter := nextCounter(r)
+	now := time.Now().Format("2006-01-02T150405")
+	baseName := fmt.Sprintf("r%d.%06d.%v", r.Revision(), counter, now)
+
+	logPath := filepath.Join(rundir, baseName+".output")
+	logf, err := os.OpenFile(logPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer logf.Close()
+
+	// run things and captures output etc
+	stR, stW, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	defer stR.Close()
+	defer stW.Close()
+
+	// run the script
+	env := os.Environ()
+	// we need to hardcode FD=3 because this is the FD after
+	// exec.Command() forked. there is no way in go currently
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
```

Comment:
Thanks for the question, I think the comment is important. I wish there was something like python subprocess `preexec_fn`